include/GafferBindings/SerialisationBinding.h

```diff
@@ -0,0 +1,187 @@
+//////////////////////////////////////////////////////////////////////////
+//
+//  Copyright (c) 2018, Image Engine Design Inc. All rights reserved.
+//
+//  Redistribution and use in source and binary forms, with or without
+//  modification, are permitted provided that the following conditions are
+//  met:
+//
+//      * Redistributions of source code must retain the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer.
+//
+//      * Redistributions in binary form must reproduce the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer in the documentation and/or other materials provided with
+//        the distribution.
+//
+//      * Neither the name of John Haddon nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#ifndef GAFFERBINDINGS_SERIALISATIONBINDING_H
+#define GAFFERBINDINGS_SERIALISATIONBINDING_H
+
+#include "boost/python.hpp"
+
+#include "GafferBindings/Serialisation.h"
+
+#include "IECorePython/RefCountedBinding.h"
+
+#include "boost/python/suite/indexing/container_utils.hpp"
+
+namespace GafferBindings
+{
+
+template<typename T, typename Base, typename TWrapper=T>
+class SerialiserClass : public IECorePython::RefCountedClass<T, Base, TWrapper>
+{
+	public :
+
+		SerialiserClass( const char *name );
+
+};
+
+template<typename WrappedType>
+class SerialiserWrapper : public IECorePython::RefCountedWrapper<WrappedType>
+{
+
+	public :
+
+		SerialiserWrapper( PyObject *self )
+			:	IECorePython::RefCountedWrapper<WrappedType>( self )
+		{
+		}
+
+		void moduleDependencies( const Gaffer::GraphComponent *graphComponent, std::set<std::string> &modules, const Serialisation &serialisation ) const override
+		{
+			if( this->isSubclassed() )
+			{
+				IECorePython::ScopedGILLock gilLock;
+				boost::python::object f = this->methodOverride( "moduleDependencies" );
```

Comment:
Should we be catching `boost::python::error_already_set` and calling `IECorePython::ExceptionAlgo::translatePythonException();` for all of these overrides?

Reply:
Good question. The short answer is that this code is basically unchanged from before, it's just been moved from SerialisationBinding.cpp and templated, so it's not newly broken or anything, and we've run with this for a long time.

The long answer is a bit more involved. Although you can see we're using `ScopedGILLock` here, several of the C++ serialisers actually assume they have the lock already, and go ahead and use boost::python anyway. They're relying on the fact that serialisers to date have been used exclusively behind the [lock in the Serialisation constructor](https://github.com/GafferHQ/gaffer/blob/master/src/GafferBindings/Serialisation.cpp#L68). You could make the argument that if we're relying on that lock on the outside, then we should do the exception handling on the outside too (alongside the lock in the Serialisation constructor).

So, I'm not sure. We should catch `boost::error_already_set` somewhere though, for sure. Belt and braces maybe? One good case for putting the handling in the overrides here is so that we don't have any bad examples for folks to follow, since every other override does need the handling.

Reply:
I lean towards doing it here to avoid bad examples. I definitely only noticed it missing because of the recent Cortex PR where we cleaned up all the other overrides.

Reply:
Sounds good - done in 9740d51.